### ISSUE TITLE
Move UI screenshots to git-lfs

### ISSUE THREAD

**Issue by mnapoli** [piwik/piwik#7726]
Storing the UI tests screenshots in a separate repository has a lot of problems, and it's even more a pain since we now use branches more. We cannot work on UI tests in pull requests (else we would have a lot of merge conflicts on the tests-screenshots submodule).

A solution to try would be [git-lfs](https://git-lfs.github.com/), which is a solution made by GitHub to store large files. It works like the files were committed in git, so it seems really simple to use. More information are also available [in the blog post](https://github.com/blog/1986-announcing-git-large-file-storage-lfs).

> Every user and organization on GitHub.com with Git LFS enabled will begin with 1 GB of free file storage and a monthly bandwidth quota of 1 GB. If your workflow requires higher quotas, you can easily purchase more storage and bandwidth for your account.

Things to test:

- does it work as we expect?
- does it behave correctly in pull requests/diffs on github, and merges in CLI/PhpStorm?
- does every contributor have to install it, even if they don't change the ui screenshots?
- get an idea of the size we would need, will we be close to the quota, much more, etc.

**Comment by mnapoli** [comment]
For the record @mattab subscribed to the beta but we need to wait to have it enabled.

**Comment by mattab** [comment]
Update: github gave us access to git-lfs. Next step is to try to use it in coming weeks

**Comment by mnapoli** [comment]
Some more info:

- seems to be supported on all platforms (Linux, Mac, Windows)
    - on Mac can be installed with `brew`, very easy, up to date
- when cloned on a system without the lfs extension, binary files contain text, for example:

    ```
    version https://git-lfs.github.com/spec/v1
    oid sha256:4396da40a63324dfe9f2b13f7ccede396d45885e35286b29ddb0df53bd5beb82
    size 231903
    ```
- on GitHub it shows as an image correctly
- diffs and pull requests work on Github (for example: https://github.com/piwik/test-git-lfs/pull/1/files)
- in PhpStorm I was able to commit the files and push without any problem. However the diff doesn't seem to work (the "before" image doesn't show) but this is not a blocker at all, everything still works anyway

It just works, this is pretty amazing.

**Comment by mnapoli** [comment]
Last thing: I have tested creating a conflict (https://github.com/piwik/test-git-lfs/pull/2) and merging the conflict in PhpStorm: no problem here I can merge it. Here is the screenshot:

![capture d ecran 2015-05-11 a 11 07 35](https://cloud.githubusercontent.com/assets/720328/7556479/26ed7844-f7ce-11e4-9624-0ebf9c3370df.png)

I think this is ready for integration in the main piwik repository.

**Comment by mnapoli** [comment]
Current status:

- @diosmosis had authentication problems when pushing, might be related to using gnome-keyring… need to see with the GitHub support and find a solution
- we need to make a list of all the tools we have that might use the git repository, and check that this change wouldn't break them

**Comment by mnapoli** [comment]
@mattab I think you are the best to do this: could you list of all the tools that use Piwik's git repository? The idea is to see if there could be some that break with using git-lfs. I think the risk is very low because when git-lfs isn't installed the images are just text files (so the repository still works).

Here is the list I came up with, please expand it:

- Travis -> git-lfs will be added in the next VM update ([the one that was rolled back](http://docs.travis-ci.com/user/build-environment-updates/2015-04-09/)), in the meantime we can install it manually
- Scrutinizer -> we don't care about UI screenshots there
- Piwik.org Git mirror -> that means that we will not have the screenshots on the mirror (or we could install git-lfs there to have them…)
- Release script & Debian release -> we don't want to deploy screenshots in the release, we'll need to remove the directory if it's not already done (e.g. removing the `tests/` directory)
- Running tests on AWS -> we'll need to update the provisioning to install git-lfs (to run UI tests)

**Comment by mattab** [comment]
@mnapoli  AFAIK you listed all possible uses there.

 * About Release scripts: it only checks the submodules that are listed here: https://github.com/piwik/piwik-package/blob/master/scripts/build-package.sh#L20 which does not include UI tests submodule (anyway the directory tests/ is deleted in the release)
 * Git mirror: we actually discontinued the git mirror in #6676 - some day we will re-create it (eg. using maybe gitlab) but for now we don't need to worry

I guess git-lfs is ready  to be used in Piwik :+1:

**Comment by mnapoli** [comment]
Awesome thanks!

**Comment by mnapoli** [comment]
Current status: blocked by https://github.com/github/git-lfs/issues/325

**Comment by mnapoli** [comment]
Previous issue was solved, currently blocked by Travis not being able to download the files:

```
Downloading tests/UI/expected-screenshots/ActionsDataTable_column_sorted.png (200.20 KB)
Username for '://': 

No output has been received in the last 10 minutes, this potentially indicates a stalled build or something wrong with the build itself.

The build has been terminated
```

https://travis-ci.org/piwik/piwik/jobs/63664099

Waiting for support help (since piwik/piwik is public it shouldn't ask for a password).

**Comment by diosmosis** [comment]
FYI, you might be able to use one of the automation users + the `$GITHUB_USER_TOKEN` environment var (just make sure it doesn't appear in git command output).

**Comment by mnapoli** [comment]
:+1: good to know, here is the answer I got by email:

> Git LFS doesn't support git:// remote endpoints. This is not a standard protocol, so we're not quite sure how to handle it.

> Also, we have an issue open to support public repositories without entering a password (or even being prompted for one).

> https://github.com/github/git-lfs/issues/243

Need to look into it.

**Comment by mattab** [comment]
maybe we can contact Travis CI support and ask if they can use something else than `git://`?

or @mnapoli what do you think is the next step?

**Comment by mnapoli** [comment]
I did two days ago, currently waiting for an answer.

**Comment by mnapoli** [comment]
Current status: we can override the remote URL with `git remote set-url` to replace it with a HTTPS URL. However git-lfs requires authentication even for public repositories (https://github.com/github/git-lfs/issues/243). This will be fixed in the [future version 0.5.2](https://github.com/github/git-lfs/issues/281#issue-73067303) so we'll have to wait for a fix.

**Comment by mnapoli** [comment]
v0.5.2 was released yesterday, need to try it out. However they didn't fix our issue so I don't have much hope, it seems we'll have to wait for [v0.5.3](https://github.com/github/git-lfs/issues/425)

**Comment by mnapoli** [comment]
Need to check this out again since https://github.com/github/git-lfs/issues/243 was closed. But we might still need to wait for v0.6.0 to be released.

**Comment by mnapoli** [comment]
Issues to look at for:

- https://github.com/github/git-lfs/issues/425
- https://github.com/github/git-lfs/issues/426

**Comment by mnapoli** [comment]
Good news, it seems to be working with v0.5.2 on travis, however it requires to have a github token in the remote URL (https://github.com/piwik/piwik/blob/git-lfs/tests/travis/install_git_lfs.sh#L7-7). So it's not usable right now.

We still have to wait for 0.6.0.

**Comment by diosmosis** [comment]
I think it's ok if we redirect output /dev/null?

**Comment by mnapoli** [comment]
But plugins do not always have a github token setup?

**Comment by diosmosis** [comment]
They should for auto updating the .travis.yml, and it can be added through generate:Travis-yml. I guess it wouldn't work for 3rd party plugins, but I don't think they matter as much.

**Comment by mnapoli** [comment]
Will try tomorrow to move all the screenshots to git lfs (in the branch) and see how it all behaves then

**Comment by diosmosis** [comment]
:+1: party time!

**Comment by quba** [comment]
@mattab, @diosmosis is it doable now?

**Comment by diosmosis** [comment]
@quba, @tsteur tried to get the PR to work but ran into some problems I think.

**Comment by tsteur** [comment]
Tried to make it work last week too and didn't work

**Comment by mattab** [comment]
Using Git LFS is awesome and makes our life so much easier. Well done guys
